sys/exec.c

```diff
@@ -1,21 +1,162 @@
+    vm_map_t* vmap = vm_map_new(PMAP_USER, 123);
+    // Note: we do not claim ownership of the map
+    set_active_vm_map(vmap);
+
+    // Iterate over prog headers
+    kprintf("[exec] ELF has %d program headers\n", eh->e_phnum);
+
+    const Elf32_Phdr* phs = (Elf32_Phdr*)(elf_image + eh->e_phoff);
+    for(uint8_t i = 0; i < eh->e_phnum; i++){
+        const Elf32_Phdr* ph = phs + i;
+        switch(ph->p_type){
+        case PT_NULL:
+        case PT_NOTE:
+        case PT_PHDR:
+        default:
+            // Ignore the section.
+            break;
+        case PT_DYNAMIC:
+        case PT_INTERP:
+            kprintf("[exec] Exec failed: ELF file requests dynamic linking"
+                    "by providing a PT_DYNAMIC and/or PT_INTERP segment.\n");
+            return -1;
+        case PT_SHLIB:
+            kprintf("[exec] Exec failed: ELF file contains a PT_SHLIB segment\n");
+            return -1;
+        case PT_LOAD:
+            kprintf("[exec] Processing a PT_LOAD segment: VirtAddr = %p, "
+                    "Offset = 0x%08x, FileSiz = 0x%08x, MemSiz = 0x%08x, "
+                    "Flags = %d\n", (void*)ph->p_vaddr, (unsigned int)ph->p_offset,
+                    (unsigned int)ph->p_filesz, (unsigned int)ph->p_memsz,
+                    (unsigned int)ph->p_flags);
+            if(ph->p_vaddr % PAGESIZE){
+                kprintf("[exec] Exec failed: Segment p_vaddr is not page alligned\n");
+                return -1;
+            }
+            vm_addr_t start = ph->p_vaddr;
+            vm_addr_t end   = roundup(ph->p_vaddr + ph->p_memsz, PAGESIZE);
+            // TODO: What if segments overlap?
+            // Temporarily permissive protection.
+            vm_map_entry_t* segment = vm_map_add_entry(vmap, start, end,
+                                    VM_PROT_READ|VM_PROT_WRITE|VM_PROT_EXEC);
+            // Allocate pages backing this segment.
+            segment->object = default_pager->pgr_alloc();
+            // Copy data into the segment
+            memcpy((uint8_t*)start, elf_image + ph->p_offset, ph->p_filesz);
+            // Zero the rest
+            if(ph->p_filesz > ph->p_memsz){
+                bzero((uint8_t*)start + ph->p_filesz, ph->p_memsz - ph->p_filesz);
+            }
+            // Apply correct permissions
+            vm_prot_t prot = VM_PROT_NONE;
+            if(ph->p_flags | PF_R) prot |= VM_PROT_READ;
+            if(ph->p_flags | PF_W) prot |= VM_PROT_WRITE;
+            if(ph->p_flags | PF_X) prot |= VM_PROT_EXEC;
+            // Note: vm_map_protect is not yet implemented, so
+            // this will have no effect as of now
+            vm_map_protect(vmap, start, end, prot);
+        }
+    }
+
+    vm_map_dump(vmap);
+
+    // We need to correct the thread structure so that it can hold the
+    // correct $gp value for this execution context.
+    thread_t* th = thread_self();
+    th->td_kctx.gp = 0;
+    th->td_kctx.ra = 0xdeadbeef;
+    th->td_kctx.pc = eh->e_entry;
```

Comment:
Note, that this is not enough. ELF files have specific requirements regarding their entry environment. For example, they need to be passed arguments list, environment variables and auxiliary vector. This should be sufficient for a very simple example, though.